File: nsinfo.py
#!/usr/bin/env python
#

"""
nsinfo.py

For a given DNS zone, print out some information about its authoritative
nameservers, specifically nameserver name, IP address, ASN originating
prefix covering the IP address, and reverse mapping of IP address to name.

Author: Shumon Huque <shuque@gmail.com>

Example run:

$ nsinfo.py complex.com
ns1.complex.com. 192.16.16.5 AS15133 ns1.edgecastdns.net.
ns1.complex.com. 2606:2800:3::5 AS15133 ns1.edgecastdns.net.
ns2.complex.com. 192.16.16.6 AS15133 ns2.edgecastdns.net.
ns2.complex.com. 2606:2800:3::6 AS15133 ns2.edgecastdns.net.
ns3.complex.com. 198.7.29.5 AS15133 ns3.edgecastdns.net.
ns3.complex.com. 2606:2800:c::5 AS15133 ns3.edgecastdns.net.
ns4.complex.com. 198.7.29.6 AS15133 ns4.edgecastdns.net.
ns4.complex.com. 2606:2800:c::6 AS15133 ns4.edgecastdns.net.

"""

import os.path
import sys
import getopt
import socket
import struct
import dns.resolver, dns.reversename

PYVERSION      = sys.version_info.major
PROGNAME       = os.path.basename(sys.argv[0])
VERSION        = "0.1"

IP2ASN_V4_SUFFIX = ".origin.asn.cymru.com."
IP2ASN_V6_SUFFIX = ".origin6.asn.cymru.com."


def usage():
    print("Usage: {} [-4|6] <domain>".format(PROGNAME))
    sys.exit(1)


def get_resolver(timeout=5, edns=False):
    """return initialized resolver object"""
    r = dns.resolver.Resolver()
    r.lifetime = timeout
    if edns:
        r.use_edns(edns=0, ednsflags=0, payload=4096)
    return r


def do_query(r, qname, qtype, qclass='IN', quiet_notfound=False):
    """Perform DNS query and return answer RRset object"""
    response = None
    try:
        answers = r.query(qname, qtype, qclass)
    except (dns.resolver.NoAnswer, dns.resolver.NXDOMAIN):
        if not quiet_notfound:
            print("{}/{}/{}: No records found.".format(qname, qtype, qclass))
    except (dns.exception.Timeout):
        print("{}/{}/{}: Query timed out.".format(qname, qtype, qclass))
    except Exception as e:
        print("{}/{}/{}: error: {}".format(qname, qtype, qclass, 
                                           type(e).__name__))
    else:
        response = answers.rrset

    return response


def reverse_octets(packedstring):
    if PYVERSION == 2:
        return ["%d" % ord(x) for x in packedstring]
    else:
        return ["%d" % x for x in packedstring]


def reverse_hexdigits(packedstring):
    if PYVERSION == 2:
        return ''.join(["%02x" % ord(x) for x in packedstring])
    else:
        return ''.join(["%02x" % x for x in packedstring])


def ip2asn(res, address):
    """
    TXT records queried return single strings of the form:
    ASN | IPprefix | CountryCode | RIR | date, e.g.
    '55 | 128.91.0.0/16 | US | arin | '
    '55 | 2607:f470::/32 | US | arin | 2008-05-01'
    """
    qname = None
    try:
        if address.find('.') != -1:
            packed = socket.inet_pton(socket.AF_INET, address)
            octetlist = reverse_octets(packed)
            qname = '.'.join(octetlist[::-1]) + IP2ASN_V4_SUFFIX
        elif address.find(':') != -1:
            packed = socket.inet_pton(socket.AF_INET6, address)
            hexlist = reverse_hexdigits(packed)
            qname = '.'.join(hexlist[::-1]) + IP2ASN_V6_SUFFIX
    except socket.error:
        pass
    if not qname:
        raise ValueError("%s isn't an IP address" % address)

    txt_rrset = do_query(res, qname, 'TXT')
    if txt_rrset:
        return "AS" + txt_rrset[0].strings[0].split('|')[0].rstrip(' ')
    else:
        return None


def ip2name(res, address):
    ptr_rrset = do_query(res, dns.reversename.from_address(address), 'PTR')
    if ptr_rrset:
        return ptr_rrset[0].target
    else:
        return None


if __name__ == '__main__':

    try:
        (options, args) = getopt.getopt(sys.argv[1:], '46')
    except getopt.GetoptError:
        usage()
    if len(args) != 1:
        usage()

    addrtypes = [ 'AAAA', 'A' ]
    for (opt, optval) in options:
        if opt == "-4":
            addrtypes = [ 'A' ]
        elif opt == "-6":
            addrtypes = [ 'AAAA' ]

    qname = args[0]

    res = get_resolver()

    ns_rrset = do_query(res, qname, 'NS')
    if not ns_rrset:
        sys.exit(1)

    hostlist = []
    for ns_rr in ns_rrset:
        for rrtype in addrtypes:
            ip_rrset = do_query(res, ns_rr.target, rrtype, quiet_notfound=True)
            if ip_rrset:
                for ip_rr in ip_rrset:
                    hostlist.append( (ns_rr.target, ip_rr.address) )

    for (ns, addr) in sorted(hostlist):
        print("{} {} {} {}".format(
            ns, addr, ip2asn(res, addr), ip2name(res, addr)))

File: nsinfo.go
package main

import (
	"fmt"
	"net"
	"os"
	"sort"
	"strings"
	"time"

	"github.com/miekg/dns"
)

// Defaults
var (
	defaultTimeout = 3
	defaultRetries = 3
)

// Globals
var resolver net.IP
var qopts = QueryOptions{rdflag: true, payload: 1460,
	timeout: time.Second * time.Duration(defaultTimeout),
	retries: defaultRetries}

//
// Print usage string
//
func usage() {
	fmt.Printf("Usage: nsinfo <zonename>\n")
}

//
// Make domain name fully qualified
//
func mkFqdn(qname string) string {
	if strings.HasSuffix(qname, ".") {
		return qname
	}
	return qname + "."
}

//
// Obtain list of IPv4 and IPv6 addresses for hostname
//
func getIPAddresses(resolver net.IP, hostname string) []net.IP {

	var ipList []net.IP
	var q Query

	var rrTypes = []uint16{
		dns.TypeAAAA,
		dns.TypeA,
	}

	for _, rrtype := range rrTypes {
		q = getQuery(hostname, rrtype, dns.ClassINET)
		response, err := sendQuery(q, resolver, qopts)
		if err != nil {
			break
		}
		for _, rr := range response.Answer {
			if rr.Header().Rrtype == rrtype {
				if rrtype == dns.TypeAAAA {
					ipList = append(ipList, rr.(*dns.AAAA).AAAA)
				} else if rrtype == dns.TypeA {
					ipList = append(ipList, rr.(*dns.A).A)
				}
			}
		}
	}

	return ipList

}

//
// Return reverse DNS lookup of given IP address
//
func reverseLookup(resolver net.IP, ipaddr net.IP) string {

	arpaname, err := dns.ReverseAddr(ipaddr.String())
	if err != nil {
		return ""
	}
	q := getQuery(arpaname, dns.TypePTR, dns.ClassINET)
	response, err := sendQuery(q, resolver, qopts)
	if err != nil || len(response.Answer) < 1 {
		return "NO-PTR"
	}
	ptrRr := response.Answer[0].(*dns.PTR)
	return ptrRr.Ptr
}

const hexDigit = "0123456789abcdef"

var ip2asnSuffixV4 = "origin.asn.cymru.com."
var ip2asnSuffixV6 = "origin6.asn.cymru.com."

func ip2asn(resolver net.IP, ip net.IP) string {

	var h1, h2 byte
	var qname string

	if ip.To4() != nil {
		// IPv4 address
		qname = fmt.Sprintf("%d.%d.%d.%d.%s",
			int(ip[3]),
			int(ip[2]),
			int(ip[1]),
			int(ip[0]),
			ip2asnSuffixV4)
	} else {
		// IPv6 address
		buf6 := make([]byte, 0, len(ip)*4)
		for i := len(ip) - 1; i >= 0; i-- {
			h1 = ip[i] & 0xf
			h2 = ip[i] >> 4
			buf6 = append(buf6, hexDigit[h1])
			buf6 = append(buf6, '.')
			buf6 = append(buf6, hexDigit[h2])
			buf6 = append(buf6, '.')
		}
		qname = string(buf6) + ip2asnSuffixV6
	}

	q := getQuery(qname, dns.TypeTXT, dns.ClassINET)
	response, _ := sendQuery(q, resolver, qopts)
	if response == nil {
		fmt.Printf("No ASN found\n")
		return ""
	}

	txtRr := response.Answer[0].(*dns.TXT)
	return "AS[" + strings.TrimSuffix(strings.Split(txtRr.Txt[0], "|")[0], " ") + "]"

}

//
// main()
//
func main() {

	var err error

	if len(os.Args) != 2 {
		usage()
		return
	}

	zone := mkFqdn(os.Args[1])
	resolver, err = getResolver()

	if err != nil {
		fmt.Printf("Error obtaining resolver address: %s", err.Error())
	}

	q := getQuery(zone, dns.TypeNS, dns.ClassINET)
	response, err := sendQuery(q, resolver, qopts)
	if err != nil {
		fmt.Printf("%s\n", err.Error())
		return
	}

	var nsRr *dns.NS
	var nsNames []string

	for _, rr := range response.Answer {
		if rr.Header().Rrtype == dns.TypeNS {
			nsRr = rr.(*dns.NS)
			nsNames = append(nsNames, nsRr.Ns)
		}
	}

	// TODO: sort in DNS canonical order
	sort.Strings(nsNames)

	var ipList []net.IP

	for _, nsName := range nsNames {
		ipList = getIPAddresses(resolver, nsName)
		for _, ip := range ipList {
			fmt.Printf("%s %s %s %s\n", nsName, ip, ip2asn(resolver, ip),
				reverseLookup(resolver, ip))
		}
	}

}
